pkg/metrics/metrics.go

```diff
@@ -9,7 +9,7 @@ import (
 var (
 	builtInCollectors    = make(map[string]InitCollector)
 	registeredCollectors = []prometheus.Collector{}
-	log                  = logger.NewLogger("metrics")
+	log                  = logger.NewLogger("raw-metrics")
```

Comment:
Should we rename this package to `collectors`. That's what they are and CRI-RM uses collectors to build metrics... just a thought.

Reply:
Yes, that'd probably be a good idea.

Reply:
But I'd like to do that in connection with other API changes there. I talked about this briefly in the last Friday chat, when I mentioned that I'd like to turn the current model inside out. Currently we 
  1. collect raw data from the kernel
  2. turn it into prometheus format and collect it
  3. try to turn it into something better suited for policies
 [4. forward raw data + chewed upon/correlated metrics to prometheus. (This still missing, it's only on the original proto branch, I'll try to clean it up and file a PR for it.)

Now the prometheus data model is not well suited / very inefficient for any type of correlation/processing other than plotting or something else a time series DB is good at.
Instead of the current model I'd like to change to this one:

  1. low-level collect data either in some resource-manager/metrics.SomeMetrics format, or something which is a natural reflection of what the kernel provides. This we already have in practice in the low-level collectors. They just convert it into the prometheus format when polled.
  2. pass events in the native format up in the stack when polled by resource-manager/metrics
  3. define an interface collector.ToMetricFamily(interface{}, []*model.MetricFamily) which all the collectors must implement (or actually I think the data types should implement them)
  4. do the conversion on-demand only when/if we proxy data to prometheus

I think this would be a much more sensible model avoiding double conversion/correlation and having to operate on data in a very unsuitable format.

Reply:
But I changed the logger source name to collectors. ;-)

Reply:
> Now the prometheus data model is not well suited / very inefficient for any type of
> correlation/processing other than plotting or something else a time series DB is good at.

For avx I still want to experiment with prometheus summary/histogram functions to see if they'd somehow help with the tagging and policy/rebalancing: for instance histograms for per-CPU activity to verify "non-AVX" cores indeed remain "zero activity" OR per-container activity to rebalance "top peak" containers... (maybe in this case separate tagging/conversion is even not needed?)

Reply:
Yes, sure. In the above model nothing prevents us, in princpiple, from declaring that the native data type for AVX usage collection is a *MetricFamily, if that turns out to offer tangible benefits. Then the above (using histrogam, etc. functions from prometheus) should become straightforward, without any intermediary steps.

However, to reiterate my point about the prometheus data being inconvenient for us, look at how even for AVX, which requires almost no correlation beyond the eventual resolution (in the resource-manager layer) from cgroup to cache.Container, we need to loop through two MetricFamilies, hashing everything by cgroup path to two different tables, just to determine how actively a cgroup has been using AVX instructions.

And this gets at least linearly worse (scaling by the number of pieces of information you want to cross-correlate) as you try to stitch back together pieces of information which was first chopped up by the low-level raw metrics collectors. It would make more sense IMO to do the necessary chopping up at the edge, on the cri-resmgr -> prometheus edge, and only if/when we proxy to prometheus.